Tasks/PWGHF/HFCandidateCreatorBPlus.cxx

```diff
@@ -0,0 +1,291 @@
+                         std::sqrt(impactParameter0.getSigmaY2()), std::sqrt(impactParameter1.getSigmaY2()),
+                         candidate.globalIndex(), track.globalIndex(), //index D0 and bachelor
+                         hfFlag);
+      } //track loop
+    }   //D0 cand loop
+  }     //process
+};      //struct
+
+/// Extends the base table with expression columns.
+struct HfCandidateCreatorBplusExpressions {
+  Spawns<aod::HfCandBPlusExt> rowCandidateBPlus;
+  void init(InitContext const&) {}
+};
+
+/// Performs MC matching.
+struct HfCandidateCreatorBplusMc {
+  Produces<aod::HfCandBPMCRec> rowMCMatchRec;
+  Produces<aod::HfCandBPMCGen> rowMCMatchGen;
+
+  void process(aod::HfCandBPlus const& candidates,
+               aod::HfCandProng2,
+               aod::BigTracksMC const& tracks,
+               aod::McParticles const& particlesMC)
+  {
+    int indexRec = -1, indexRecD0 = -1;
+    int8_t signB = 0, signD0 = 0;
+    int8_t flag = 0;
+    int kD0pdg = pdg::Code::kD0;
+
+    // Match reconstructed candidates.
+    for (auto& candidate : candidates) {
+      //Printf("New rec. candidate");
+
+      flag = 0;
+      auto candDaughterD0 = candidate.index0_as<aod::HfCandProng2>();
+      auto arrayDaughtersD0 = array{candDaughterD0.index0_as<aod::BigTracksMC>(), candDaughterD0.index1_as<aod::BigTracksMC>()};
+      auto arrayDaughters = array{candidate.index1_as<aod::BigTracksMC>(), candDaughterD0.index0_as<aod::BigTracksMC>(), candDaughterD0.index1_as<aod::BigTracksMC>()};
+
+      // B± → D0bar(D0) π± → (K± π∓) π±
+      //Printf("Checking B± → D0(bar) π±");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kBPlus, array{+kPiPlus, +kKPlus, -kPiPlus}, true, &signB, 2);
+      indexRecD0 = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughtersD0, pdg::Code::kD0, array{-kKPlus, +kPiPlus}, true, &signD0, 1);
+
+      if (indexRecD0 > -1 && indexRec > -1) {
+        flag = signB * (1 << hf_cand_bplus::DecayType::BPlusToD0Pi);
+      }
+      rowMCMatchRec(flag);
+    }
+
+    // Match generated particles.
+    for (auto& particle : particlesMC) {
+      //Printf("New gen. candidate");
+      flag = 0;
+      signB = 0;
+      signD0 = 0;
+      int indexGenD0 = -1;
+
+      // B± → D0bar(D0) π± → (K± π∓) π±
+      //Printf("Checking B± → D0(bar) π±");
+      std::vector<int> arrayDaughterB;
+      if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kBPlus, array{-kD0pdg, +kPiPlus}, true, &signB, 1, &arrayDaughterB)) {
+        // D0(bar) → π± K∓
+        //Printf("Checking D0(bar) → π± K∓");
+        for (auto iD : arrayDaughterB) {
+          auto candDaughterMC = particlesMC.iteratorAt(iD);
+          if (std::abs(candDaughterMC.pdgCode()) == kD0pdg) {
+            indexGenD0 = RecoDecay::isMatchedMCGen(particlesMC, candDaughterMC, pdg::Code::kD0, array{-kKPlus, +kPiPlus}, true, &signD0, 1);
+          }
+        }
```

Comment:
In order to make sure the B candidates are matched with the desired decay channel we check the PDG of decay daughters. Similarly for D0 candidates as well. I havent checked if not doing this cross check would work and give correct matching. I can do this check and then modify the code if needed.

Reply:
The matching is done in `RecoDecay::isMatchedMCGen`.
You seem to be looping over the matched daughters of B+ to find the D0bar and run the matching for it.
My point is:
* The daughters of MC particles should be ordered from the way they were stored in the table. This order should be preserved in `arrayDaughterB` so you should know which index is D0 and which one is pion. Therfore you should not need to loop over them and check their PDG code.
* The entire matching of D0 candidates seems redundant because it is already done in the 2-prong candidate creator. Why would you need to do it again here?